### PR TITLE
UBERF-4195: fix query after applying viewOptions

### PR DESCRIPTION
**Pull request by ThetaDR**
# Contribution checklist

## Checklist

* [ ] - UI test added to added/changed functionality?
* [ ] - Screenshot is added to PR if applicable ?
* [x] - Does the code work? Check whether function and logic are correct.
* [ ] - Does Changelog.md is updated with changes?
* [ ] - Does the translations are up to date?
* [ ] - Does it well tested?
* [x] - Tested for Chrome.
* [ ] - Tested for Safari.
* [x] - Go through the changed code looking for typos, TODOs, commented LOCs, debugging pieces of code, etc.
* [ ] - Rebase your branch onto master and upstream branch
* [ ] - Is there any redundant or duplicate code?
* [ ] - Are required links are linked to PR?
* [ ] - Does new code is well documented ?

## Related issues

A list of closed updated issues

### CODE REVIEW

**Review by BykhovDenis — CHANGES_REQUESTED** [review]
This will break viewOptions, can you give an example? (If it necessary we should find another way to fix it, probably we need method in core to merge queries)

**Comment by ThetaDR** [comment]
@BykhovDenis long story short. It replaces spaces that we are using in filter with spaces from viewOptions (for non-archive)

More info:
1. Company1 has two vacancies (vac1 and vac2)
Company2 has vac3
2. All of them have applicants 
3. Vac1 is archived
4. User is trying to filter applicants via vacancy with company == Company1. And viewOption for hiding archived vacancies is on.

So it will override spaces (vac1 and vac2) with spaces that are not archived (vac1 and vac3)

**Comment by ThetaDR** [comment]
@BykhovDenis PTAL, made a function for queries merging

File: packages/core/src/utils.ts [hunk 1]
```diff
@@ -331,3 +331,38 @@ export class RateLimitter {
...
   }
 }
+
+export function mergeQueries<T extends Doc> (query1: DocumentQuery<T>, query2: DocumentQuery<T>): DocumentQuery<T> {
+  const q = { ...query1 }
+  for (const k in query2) {
+    if (
+      typeof query2[k] === 'object' &&
+      typeof query1[k] === 'object' &&
+      Object.keys(query2[k]).every((t) => t.startsWith('$')) &&
```
Comment (haiodo) [inline packages/core/src/utils.ts]: query1[k] could be non object and we still should merge.
Please use isPredicate(query2[key])
query2[k] could also be non object and we should merge.

```ts
query1 = {
    space: prj1._id
}
query2 = {
   space: {$in: [prj1._id, prj2._id]
}
```
and wise verse

File: plugins/task-resources/src/components/kanban/KanbanView.svelte [hunk 2]
```diff
@@ -87,7 +88,7 @@
   }
 
   let resultQuery: DocumentQuery<any> = { ...query }
-  $: getResultQuery(query, viewOptionsConfig, viewOptions).then((p) => (resultQuery = { ...p, ...query }))
+  $: getResultQuery(query, viewOptionsConfig, viewOptions).then((p) => (resultQuery = mergeQueries(p, query)))
```
Comment (haiodo) [inline plugins/task-resources/src/components/kanban/KanbanView.svelte:91]: We should reverse order here?
```suggestion
  $: getResultQuery(query, viewOptionsConfig, viewOptions).then((p) => (resultQuery = mergeQueries(query, p)))
```

So we will merge into base one, filter ones.?

File: plugins/view-resources/src/components/list/List.svelte [hunk 3]
```diff
@@ -59,7 +68,7 @@
 
   let resultQuery: DocumentQuery<Doc> = query
   $: getResultQuery(query, viewOptionsConfig, viewOptions).then((p) => {
-    resultQuery = { ...query, ...p }
+    resultQuery = mergeQueries(p, query) as DocumentQuery<Doc>
```
Comment (haiodo) [inline plugins/view-resources/src/components/list/List.svelte:71]: And reverse here or not?

File: plugins/task-resources/src/components/kanban/KanbanView.svelte (hunk 2 above)
Comment (ThetaDR) [reply plugins/task-resources/src/components/kanban/KanbanView.svelte:91]: It should not care about the order of arguments, so supposed to give the same answer no matter of the order

File: packages/core/src/utils.ts [hunk 4]
```diff
@@ -331,3 +332,44 @@ export class RateLimitter {
...
+      for (const x in toIterate) {
+        if (['$in', '$nin'].includes(x)) {
+          const q1Arr = isPredicate(query1[k]) ? query1[k][x] : [query1[k]]
+          const q2Arr = isPredicate(query2[k]) ? query2[k][x] : [query2[k]]
+          if (x === '$in') {
+            const intersection =
+              q1Arr.length - q2Arr.length < 0
+                ? q2Arr.filter((c: any) => q1Arr.includes(c))
+                : q1Arr.filter((c: any) => q2Arr.includes(c))
+            Object.assign(q, { [k]: { $in: intersection } })
```
Comment (haiodo) [inline packages/core/src/utils.ts]: $nin is lost

File: packages/core/src/utils.ts [hunk 5]
```diff
@@ -331,3 +332,44 @@ export class RateLimitter {
...
+
+export function mergeQueries<T extends Doc> (query1: DocumentQuery<T>, query2: DocumentQuery<T>): DocumentQuery<T> {
+  const q = Object.assign({}, query1)
+  for (const k in query2) {
+    if (!Object.keys(query1).includes(k)) {
+      Object.assign(q, { [k]: query2[k] })
+      continue
+    }
+    if (isPredicate(query2[k]) || isPredicate(query1[k])) {
+      const toIterate = isPredicate(query2[k]) ? query2[k] : query1[k]
```
Comment (haiodo) [inline packages/core/src/utils.ts:345]:
```ts

const aIn = ('$in' in query1[k] ? query1[k].$in: undefined) ?? (typeof query1[k] !== 'object' && query1[k] !== undefined: [query1[k]] : [])
const aNIn = ('$nin' in query1[k] ? query1[k].$nin: undefined) ?? (typeof query1[k] === 'object' && query1[k].$ne !== undefined: [query1[k].$ne] : [])

const bIn = ('$in' in query1[k] ? query1[k].$in: undefined) ?? (typeof query1[k] !== 'object' && query1[k] !== undefined: [query1[k]] : [])
const bNIn = ('$nin' in query1[k] ? query1[k].$nin: undefined) ?? (typeof query1[k] === 'object' && query1[k].$ne !== undefined: [query1[k].$ne] : [])

const finalIn = aIn.length - bIn.length < 0 ? bIn.filter((c: any) => aIn.includes(c)) : aIn.filter((c: any) => bIn.includes(c))
            Object.assign(q, { [k]: { $in: intersection } })
const finalNin = Array.from(new Set([...aNIn, ... bNIn]))

if( finalin.length === 1 && finalNIn.length === 0 ) {return aIn[0]}
if( finalIn.length === 0 && finalNin.length === 1) {return {$ne: finalNIn[0]}}
const res: any = {}
if( finalIn.length > 0) {res.$in = finalIn}
if( finalNIn.length > 0) {res.$nin = finalNIn}
return res

```

**Review by haiodo — APPROVED** [review]

**Review by BykhovDenis — APPROVED** [review]

**Pull request merged**